### ISSUE TITLE
Adding some new patchers, updating a bunch of old ones

### ISSUE THREAD

**Issue by Infernio** [wrye-bash/wrye-bash#461]
Collecting my ongoing patcher updates and enhancements here.

## New Patchers

1. **Import Keywords**
 As it stands, keywords (a combination of the `KWDA` and `KSIZ` subrecords) occur in plenty of places in Skyrim-based records (search `MelKeywords` for an overview), but Wrye Bash can't actually carry changes to these forward. As mentioned in #305, a proper keywords patcher is needed.

1. **Import Object Bounds**
 Same story as with Import Keywords, `OBND - Object Bounds` are *very* common subrecords, and can't currently be patched.

1. **Import Text**
 Would patch DESC subrecords and the DNAM description subrecord for MGEF in Skyrim. Intuition would be to patch descriptions of armor, weapons, spells , etc and text contents of things like books. Applies to all games but Fallout 4 (due to missing record definitions). Needs a dict mapping record types to the tuple it should patch, because calling the text in a book a 'description' is silly. Open to a better name for this.

## Updated Patchers

1. **Contents Checker**
 Previously supported only Oblivion, will now be available for FO3, FNV, Skyrim, Enderal and SSE as well. Information about the allowed entry types comes directly from xEdit, which happened to line up with valda's version of the Contents Checker for FO3 and FNV as well.

1. **Import Actors**
 Previously supported only Oblivion, will now be available for FO3, FNV, Skyrim, Enderal and SSE as well. Note that `Creatures.Blood` is not available for Skyrim, because creatures don't exist in the newer games.

1. **Import Actors: AI Packages**
 Previously supported only Oblivion, will now be available for FO3, FNV, Skyrim, Enderal and SSE as well.

1. **Import Actors: Animations**
 Previously supported only Oblivion, will now be available for FO3 and FNV as well. Skyrim does not have an equivalent to the patched subrecord type.

1. **Import Actors: Death Items**
 Previously supported only Oblivion, will now be available for FO3, FNV, Skyrim, Enderal and SSE as well.

1. **Import Actors: Spells**
 Previously supported only Oblivion, will now be available for FO3, FNV, Skyrim, Enderal and SSE as well. Note that valda's version did not have this patcher, but the record definitions to support it are there - Fallout games just call these 'actor effects' instead of 'spells'.

1. **Import Destructible**
 Previously supported only FO3 and FNV, will now be available for Skyrim, Enderal and SSE as well. Fairly useless, since there is only [one mod](https://www.nexusmods.com/skyrimspecialedition/mods/28291) I can find for Skyrim that makes use of destructibles, so this is more about code quality (rule of three and all that).

1. **Import NPC Faces**
 Previously supported only Oblivion, will now be available for FO3 and FNV as well. Not sure if this one can go any further.

1. **Import Scripts**
 Previously supported only Oblivion, will now be available for FO3 and FNV as well. We're also able to support more types than in valda's version here, based on xEdit and our own record definitions. Brings us a step closer to valda's version in terms of available patchers.

1. **Import Spell Stats**
 Previously supported only Oblivion, will now be available for FO3, FNV, Skyrim, Enderal and SSE as well. Another patcher that was not present in valda's version, but works fine with the FO3/FNV record definitions.

1. **Tweak Settings**
 Adds more tweaks for Skyrim, Enderal and SSE. Includes @BeermotorWB's new tweaks, somewhat edited, and a few other ones I ported from Oblivion.

## Upcoming in `inf-312-parsers-abc`
1. **Import Factions**
 Previously supported only Oblivion, FO3 and FNV, will now be available for Skyrim, Enderal and SSE as well. Still has an ancient bug where it doesn't actually carry any deletions of factions forward, which I want to fix before merging the branch.

1. **Import Relations**
 Previously supported only Oblivion and (incorrectly) FO3 and FNV, will now be available for Skyrim, Enderal and SSE as well. Also had the same bug as above regarding deletions, which I've already fixed. It also wasn't properly updated for FO3 and FNV, being the cause of the `groupCombatReaction` crash.

## Future Work
1. **Import Actors: Perks**
 Could be done now, but would involve a lot of copy-pasting from `Import Actors: AI Packages`, which is not going to happen. So will have to wait until 308.

1. **Race Records**
 Complete nightmare. See `inf-wip` for my half-baked attempt at splitting out the horror that is 'googly eye fixing' and 'eye and hair sorting' into a new patcher.

1. **Update References**
 I'll be honest, I have no clue how this thing is supposed to work. Half of it is commented out, which doesn't exactly inspire confidence.

Related to #151 and #468.

**Comment by Infernio** [comment]
Turns out that this is indeed a `_SimpleImporter`, meaning that its implementation is so simple, it only takes ~30 lines of code (including the new Bash Tag). Moving to 307 since this turned out to be so simple.

Also helps that I don't have to worry about CBash 🎉

**Comment by Utumno** [comment]
feel free to add docs to that _Simple Importer of mine :p - actually this
was a blind refactoring of (lots of) duplicate code

**Comment by Utumno** [comment]
Quick note - nightly Pbash on oblivion fails

<details><summary>Traceback</summary><p>

```
Oblivion.esm: CLAS.DATA: Expected one of sizes [52, 42], but got 48
Error loading 'CLAS' record and/or subrecord: 00034521
  eid = u'MarauderBoss'
  subrecord = 'DATA'
  subrecord size = 48
  file pos = 39332
Error in Oblivion.esm
parsers.py 4005 load:  
Traceback (most recent call last):
  File "bash\parsers.py", line 3999, in load
    self.tops[label].load(ins, do_unpack and (topClass != MobBase))
  File "bash\record_groups.py", line 83, in load
    ins.tell() + self.size - self.header.__class__.rec_header_size)
  File "bash\record_groups.py", line 184, in loadData
    record = recClass(header,ins,True)
  File "bash\brec.py", line 2358, in __init__
    self.__class__.melSet.initRecord(self, header, ins, do_unpack)
  File "bash\brec.py", line 1612, in initRecord
    MreRecord.__init__(record, header, ins, do_unpack)
  File "bash\brec.py", line 2149, in __init__
    if ins: self.load(ins, do_unpack)
  File "bash\brec.py", line 2213, in load
    self.loadData(ins,inPos+self.size)
  File "bash\brec.py", line 2367, in loadData
    self.__class__.melSet.loadData(self, ins, endPos)
  File "bash\brec.py", line 1643, in loadData
    self._handle_load_error(error, record, ins, sub_type, sub_size)
  File "bash\brec.py", line 1634, in loadData
    read_id_prefix + sub_type)
  File "bash\game\oblivion\records.py", line 566, in loadData
    raise ModSizeError(ins.inName, readId, (52, 42), size_)
ModSizeError: Oblivion.esm: CLAS.DATA: Expected one of sizes [52, 42], but got 48

Oblivion.esm: CLAS.DATA: Expected one of sizes [52, 42], but got 48
Error loading 'CLAS' record and/or subrecord: 00034521
  eid = u'MarauderBoss'
  subrecord = 'DATA'
  subrecord size = 48
  file pos = 39332
Error in Oblivion.esm
parsers.py 4005 load:  
Traceback (most recent call last):
  File "bash\parsers.py", line 3999, in load
    self.tops[label].load(ins, do_unpack and (topClass != MobBase))
  File "bash\record_groups.py", line 83, in load
    ins.tell() + self.size - self.header.__class__.rec_header_size)
  File "bash\record_groups.py", line 184, in loadData
    record = recClass(header,ins,True)
  File "bash\brec.py", line 2358, in __init__
    self.__class__.melSet.initRecord(self, header, ins, do_unpack)
  File "bash\brec.py", line 1612, in initRecord
    MreRecord.__init__(record, header, ins, do_unpack)
  File "bash\brec.py", line 2149, in __init__
    if ins: self.load(ins, do_unpack)
  File "bash\brec.py", line 2213, in load
    self.loadData(ins,inPos+self.size)
  File "bash\brec.py", line 2367, in loadData
    self.__class__.melSet.loadData(self, ins, endPos)
  File "bash\brec.py", line 1643, in loadData
    self._handle_load_error(error, record, ins, sub_type, sub_size)
  File "bash\brec.py", line 1634, in loadData
    read_id_prefix + sub_type)
  File "bash\game\oblivion\records.py", line 566, in loadData
    raise ModSizeError(ins.inName, readId, (52, 42), size_)
ModSizeError: Oblivion.esm: CLAS.DATA: Expected one of sizes [52, 42], but got 48

races_multitweaks.py  871 scanModFile: No right and/or no left eye recorded in race Necro Spirit Servent, from mod Windfall.esm
MERGE/SCAN ERROR: Unofficial Oblivion Patch.esp
Traceback (most recent call last):
  File "bash\balt.py", line 454, in <lambda>
    if onButClick: self.Bind(wx.EVT_BUTTON, lambda __event: onButClick())
  File "bash\balt.py", line 1798, in _conversation_wrapper
    return func(*args, **kwargs)
  File "bash\basher\patcher_dialog.py", line 211, in PatchExecute
    patchFile.scanLoadMods(SubProgress(progress,0.2,0.8)) #try to speed this up!
  File "bash\patcher\patch_files.py", line 253, in scanLoadMods
    patcher.scanModFile(modFile,nullProgress)
  File "bash\patcher\patchers\special.py", line 150, in scanModFile
    levLists[listId].mergeOverLast = True
KeyError: (bolt.Path(u'Oblivion.esm'), 247796)
```
</p></details>

Both the ModSizeError and the MERGE/SCAN ERROR: Unofficial Oblivion Patch.esp do not happen on current dev, second one crashes the patch however

Will try and pin this down if you don't get there first but posting my PBash config here for reference - I am building a huge patch to test my branch, which won't build as my branch sits on top of nightly (Oblivion is ideal as it contains all the patchers)

[Bash Patch Config.zip](https://github.com/wrye-bash/wrye-bash/files/3840476/Bash.Patch.Config.zip)

**Comment by Infernio** [comment]
See MreClas for Oblivion, specifically the subclass that implements the DATA subrecord. I tightened an `else` in there, so tracebacks like this are expected. The way to fix it is to add an `elif` branch that properly handles CLAS/DATA of size 48.

**Comment by Utumno** [comment]
Thanks for fixing the CLAS traceback (a couple moths ago) - the second one happily went away on its own accord - however on the tip of 312-patchers I have a new one that is bugging me

```
C:\Dropbox\eclipse_workspaces\python\venv\wxpython-3.2-windows\Scripts\python.exe "C:/Dropbox/eclipse_workspaces/python/wrye-bash/Mopy/Wrye Bash Launcher.pyw"
testing UAC
Traceback (most recent call last):
  File "bash\gui\events.py", line 167, in _post
    result = listener(*listener_args)
  File "bash\balt.py", line 1522, in _conversation_wrapper
    return func(*args, **kwargs)
  File "bash\basher\patcher_dialog.py", line 202, in PatchExecute
    patchFile.buildPatch(log,SubProgress(progress,0.8,0.9))#no speeding needed/really possible (less than 1/4 second even with large LO)
  File "bash\patcher\patch_files.py", line 334, in buildPatch
    patcher.buildPatch(log,SubProgress(subProgress,index))
  File "bash\patcher\patchers\multitweak_settings.py", line 242, in buildPatch
    tweak.buildPatch(self.patchFile,keep,log)
  File "bash\patcher\patchers\multitweak_settings.py", line 42, in buildPatch
    for record in patchFile.GLOB.records:
AttributeError: 'MobBase' object has no attribute 'records'
```

On nightly builds ok. I am trying to debug it currently but if I don't make it will ping (belongs to #312 no doubt)

**Comment by Infernio** [comment]
Yup, I experienced that too. It seems to now load GLOB/GMST as MobBase instead of MobObjects for some reason...

**Comment by Utumno** [comment]
Well pretty big refactoring - but after dust settles - and thanks to your bottom up records refactoring - the only dark corners that will remain is BAIN and of course cint. A long way from 304.4 - we may get more people actually capable of contributing :)

**Comment by Utumno** [comment]
Just to say I fixed - will leave groking for another time 797f28e7552945f735b94a2a74daf79320810e4c
